webpack/assets/javascripts/react_app/redux/middlewares/IntervalMiddleware/IntervalMiddleware.js

```diff
@@ -0,0 +1,48 @@
+import { omit } from 'lodash';
+import { STOP_INTERVAL, DEFAULT_INTERVAL } from './IntervalConstants';
+import { selectDoesIntervalExist, selectIntervalID } from './IntervalSelectors';
+import {
+  registeredIntervalException,
+  unregisteredIntervalException,
+} from './IntervalHelpers';
+import { startInterval as startIntervalAction } from './IntervalActions';
+import { whenDocumentIsVisible } from '../common/helpers';
+
+export const IntervalMiddleware = store => next => action => {
+  const { type, key, interval } = action;
+  const state = store.getState();
+  /**
+    for the action to run multiple times
+    without getting into an endless loop in this middleware.
+  */
+
+  const modifiedAction = omit(action, ['interval']);
+  const dispatchModifiedAction = () => store.dispatch(modifiedAction);
+
+  if (interval) {
+    if (selectDoesIntervalExist(state, key)) {
+      throw registeredIntervalException(key);
+    }
+
+    dispatchModifiedAction(); // force the action to run for the first time.
+    const delay =
+      typeof interval === 'number'
+        ? interval
+        : process.env.DEFAULT_INTERVAL || DEFAULT_INTERVAL;
+    const intervalFunc = () => whenDocumentIsVisible(dispatchModifiedAction);
+    const intervalID = setInterval(intervalFunc, delay);
+    return store.dispatch(startIntervalAction(key, intervalID));
+  }
+
+  if (type === STOP_INTERVAL) {
+    if (!selectDoesIntervalExist(state, key)) {
+      throw unregisteredIntervalException(key);
+    }
+    const intervalID = selectIntervalID(state, key);
+    clearInterval(intervalID);
+  }
+
+  return next(action);
```

Comment:
1. How about passing the  `modifiedAction` to the next middleware ? the interval key is useless  
2. in case  that the action is an API like this one:

```js
 ({
  type: API_OPERATIONS.GET,
  key: KEY,
  url,
  payload: { id },
  interval: 3000
});
```
`next` will invoke the next middleware which is the APIMiddleware and by that, it will be invoked again

Reply:
Though if it will have the `interval` key it will never get here

Reply:
The APImiddleware doesn’t filtered by the interval key, it wrapped by a condition for OPERATIONS.GET type, are you sure it won’t catch it after the interval middleware ?

Reply:
This is the flow from my understanding:

The middlewares call order: 
1. thunk
2. IntervalMiddleware
3. APIMiddleware

inside the IntervalMiddleware:
1. Received an action with the `interval` key.
2. create a modified action without the `interval` key.
3. "kills" the original flow of the action.
4. dispatch the modified action to the beginning of the middleware chain - to `thunk` in our case.
5. after that it will pass in the `IntervalMIddleware` again but this time without the `interval` key which means that it will pass to the `APIMiddleware` and do what it should.

Reply:
Thanks! I missed the `return` :man_facepalming: